sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/TestPubsubSignal.java

```diff
@@ -0,0 +1,284 @@
+   * {@link #waitForSuccess(Duration)} will unblock.
+   */
+  public <T> PTransform<PCollection<? extends T>, POutput> signalSuccessWhen(
+      Coder<T> coder,
+      SerializableFunction<Set<T>, Boolean> successPredicate) {
+
+    return new PublishSuccessWhen<>(coder, successPredicate, resultTopicPath);
+  }
+
+  /**
+   * Wait for a success signal for {@code duration}.
+   */
+  public void waitForSuccess(Duration duration) throws IOException {
+    SubscriptionPath resultSubscriptionPath = new SubscriptionPath(String.format(
+        SUBSCRIPTION_FORMAT,
+        pipelineOptions.getProject(),
+        "subscription-" + String.valueOf(ThreadLocalRandom.current().nextLong())));
+
+    pubsub.createSubscription(
+        new TopicPath(resultTopicPath),
+        resultSubscriptionPath,
+        (int) duration.getStandardSeconds());
+
+    String result = pollForResultForDuration(resultSubscriptionPath, duration);
+
+    if (!"SUCCESS".equals(result)) {
+      throw new AssertionError(result);
+    }
+  }
+
+  private String pollForResultForDuration(
+      SubscriptionPath resultSubscriptionPath,
+      Duration duration) throws IOException {
+
+    List<PubsubClient.IncomingMessage> result = null;
+    DateTime endPolling = DateTime.now().plus(duration.getMillis());
+
+    do {
+      try {
+        result = pubsub.pull(DateTime.now().getMillis(), resultSubscriptionPath, 1, false);
+        pubsub.acknowledge(
+            resultSubscriptionPath, result.stream().map(m -> m.ackId).collect(toList()));
+        break;
+      } catch (StatusRuntimeException e) {
+        System.out.println("Error while polling for result: " + e.getStatus());
+        sleep(500);
+      }
+    } while (DateTime.now().isBefore(endPolling));
+
+    if (result == null) {
+      throw new AssertionError(
+          "Did not receive success in " + duration.getStandardSeconds() + "s");
+    }
+
+    return new String(result.get(0).elementBytes, UTF_8);
+  }
+
+  private void sleep(long t) {
+    try {
+      Thread.sleep(t);
+    } catch (InterruptedException ex) {
+      throw new RuntimeException(ex);
+    }
+  }
+
+  /**
+   * {@link PTransform} that for validates whether elements seen so far match success criteria.
+   */
+  static class PublishSuccessWhen<T> extends PTransform<PCollection<? extends T>, POutput> {
+    private Coder<T> coder;
+    private SerializableFunction<Set<T>, Boolean> successPredicate;
+    private String resultTopicPath;
+
+    PublishSuccessWhen(
+        Coder<T> coder,
+        SerializableFunction<Set<T>, Boolean> successPredicate,
+        String resultTopicPath) {
+
+      this.coder = coder;
+      this.successPredicate = successPredicate;
+      this.resultTopicPath = resultTopicPath;
+    }
+
+    @Override
+    public POutput expand(PCollection<? extends T> input) {
+      return
+          input
+              // assign a dummy key, need this to keep all events in the same state cell
+              .apply(WithKeys.of("dummyKey"))
```

Comment:
What about windowing? You may also want to `Window.into(GlobalWindow)`. I think also a good amount of this logic exists in the `PAssert` implementation.